ocaml/xapi-cli-protocol/cli_protocol.ml

```diff
@@ -213,7 +213,7 @@ let marshal_message = function
   | Blob x     -> marshal_int 11 ^ (marshal_blob_header x)
 
 let write_string (fd: Unix.file_descr) buf =
-  Unixext.really_write fd buf 0 (String.length buf)
+  Unix.write fd buf 0 (String.length buf) |> ignore
```

Comment:
`Return the number of bytes actually written. write repeats the writing operation until all bytes have been written or an error occurs.`

Reply:
I suspect that the `really_write` function comes from https://ocaml.github.io/ocamlunix/ocamlunix.html (Section 5.7). Has anything changed in `Unix.write` such that `really_write` isn't useful anymore?

Reply:
Not sure, we discussed with @mseri a while ago that `Unix.write` already does what we need, it writes everything according to the docs, unlike the `write` Linux command. So the it was assumed that `Unix.write` behaves like `write(2)`, and that's why `really_write` was created.

Reply:
See that link I posted. `Unix.write` does repeated `write()` system calls, but pipes and sockets are a bit special:

> The function write of the Unix module iterates the system call write until all the requested bytes are effectively written.
> `val write : file_descr -> string -> int -> int -> int`
> However, when the descriptor is a pipe (or a socket, see chapter 6), writes may block and the system call write may be interrupted by a signal. In this case the OCaml call to Unix.write is interrupted and the error EINTR is raised. The problem is that some of the data may already have been written by a previous system call to write but the actual size that was transferred is unknown and lost. This renders the function  write of the Unix module useless in the presence of signals.

Reply:
We have actually replaced `really_write` with `Unix.write`:
https://github.com/xapi-project/stdext/blob/001b4b7386e7a411de307825a19b3fe4c51a2770/lib/xapi-stdext-unix/unixext.ml#L448

Reply:
😮

Reply:
But as you said, the implementation of `really_write` was equivalent to what `Unix.write` does, because it also just made repeated calls, and did not check the EINTR signal, so replacing it is probably harmless:
https://github.com/xapi-project/stdext/commit/1d501c6d85988019a96a5f219360a6f8acac2a60#diff-b9e2b89466fcaa1db5f80ce7b15fdf62L465
```difff
 (* Ideally, really_write would be implemented with optional arguments ?(off=0) ?(len=String.length string) *)
 let really_write_string fd string =
-  really_write fd string 0 (String.length string)
+  let payload = Bytes.unsafe_of_string string in
+  Unix.write fd payload 0 (Bytes.length payload) |> ignore
```

Reply:
The original implementation was:
```
let really_write fd string off n =
  let written = ref 0 in
  while !written < n
  do
    let wr = Unix.write fd string (off + !written) (n - !written) in
    written := wr + !written
  done
```
...which doesn't actually handle EINTR at all, so doesn't make much sense...

Reply:
A better version would use `Unix.single_write`.

Reply:
@jonludlam 🔼

Reply:
Yeah. I think it's fine to replace `really_write` with `Unix.write` though, because their behaviour is the same IMO :) - they both behave incorrectly in case of EINTR

Reply:
As on that website:
```
let rec restart_on_EINTR f x =
  try f x with Unix_error (EINTR, _, _) -> restart_on_EINTR f x
in
let rec really_write fd buffer offset len =
  let n = restart_on_EINTR (single_write fd buffer offset) len in
  if n < len then really_write fd buffer (offset + n) (len - n);;
```

Reply:
Well, if our `Unixext.really_write` was incorrect, than why don't we just fix it rather than replacing all its users with `Unix.write`?

Reply:
I think we've done a `really_write` -> `Unix.write` replacement in a past PR already, so we should undo that too, not sure which one, maybe @mseri remembers

Reply:
@robhoes thanks for the rationale behind `really_write`! When I did deprecate it in stdext nobody could tell me why we had that broken implementation, I'm happy to fix it as you suggest and revert the deprecation and changes.

I think we can figure out where we did the change and revert it once we fix the `really_write`, it's likely one of the handful PR on moving to safe-strings, and in any case it was equivalent to what it had done for the last 10 years so it should not be a problem if we miss one.

Let's not merge this yet, I'll revert the Unixext changes and then send a PR to fix Stdext